src/esp/gfx/PbrShader.cpp

```diff
@@ -0,0 +1,484 @@
+               Mn::GL::Extensions::ARB::explicit_attrib_location>(glVersion))
+#endif
+  {
+    bindAttributeLocation(Position::Location, "vertexPosition");
+    if (lightCount) {
+      bindAttributeLocation(Normal::Location, "vertexNormal");
+    }
+    if (flags_ & (Flag::NormalTexture | Flag::PrecomputedTangent) &&
+        lightCount) {
+      bindAttributeLocation(Tangent4::Location, "vertexTangent");
+    }
+    if (isTextured()) {
+      bindAttributeLocation(TextureCoordinates::Location, "vertexTexCoord");
+    }
+  }
+
+  // set texture binding points in the shader;
+  // see PBR vertex, fragment shader code for details
+  if (flags_ & Flag::BaseColorTexture) {
+    setUniform(uniformLocation("baseColorTexture"), TextureUnit::BaseColor);
+  }
+  if (flags_ & Flag::RoughnessTexture) {
+    setUniform(uniformLocation("roughnessTexture"), TextureUnit::Roughness);
+  }
+  if (flags_ & Flag::MetallicTexture) {
+    setUniform(uniformLocation("metallicTexture"), TextureUnit::Metallic);
+  }
+  if (flags_ & Flag::NormalTexture) {
+    setUniform(uniformLocation("normalTexture"), TextureUnit::Normal);
+  }
+  if ((flags_ & Flag::NoneRoughnessMetallicTexture) ||
+      flags_ & Flag::OcclusionRoughnessMetallicTexture) {
+    setUniform(uniformLocation("combinedTexture"), TextureUnit::Combined);
+  }
+
+  // cache the uniform locations
+  modelviewMatrixUniform_ = uniformLocation("ModelViewMatrix");
+  normalMatrixUniform_ = uniformLocation("NormalMatrix");
+  mvpMatrixUniform_ = uniformLocation("MVP");
+  if (flags_ & Flag::ObjectId) {
+    objectIdUniform_ = uniformLocation("objectId");
+  }
+  if (flags_ & Flag::TextureTransformation) {
+    textureMatrixUniform_ = uniformLocation("textureMatrix");
+  }
+
+  // materials
+  baseColorUniform_ = uniformLocation("Material.baseColor");
+  roughnessUniform_ = uniformLocation("Material.roughness");
+  metallicUniform_ = uniformLocation("Material.metallic");
+
+  // lights
+  if (lightCount_) {
+    lightRangesUniform_ = uniformLocation("LightRanges");
+    lightColorsUniform_ = uniformLocation("LightColors");
+    lightDirectionsUniform_ = uniformLocation("LightDirections");
+
+    if (flags_ & Flag::NormalTexture && flags_ & Flag::NormalTextureScale) {
+      normalTextureScaleUniform_ = uniformLocation("normalTextureScale");
+    }
+  }
+}
+
+// Note: the texture binding points are explicitly specified above.
+// Cannot use "explicit uniform location" directly in shader since
+// it requires GL4.3 (We stick to GL4.1 for MacOS).
+PbrShader& PbrShader::bindBaseColorTexture(Mn::GL::Texture2D* texture) {
+  CORRADE_ASSERT(flags_ & Flag::BaseColorTexture,
+                 "PbrShader::bindBaseColorTexture: the shader was not "
+                 "created with BaseColor texture enabled",
+                 *this);
+  if (lightCount_ && texture) {
+    texture->bind(TextureUnit::BaseColor);
+  }
+  return *this;
+}
+
+PbrShader& PbrShader::bindRoughnessTexture(Mn::GL::Texture2D* texture) {
```

Comment:
Hmm, why don't all those take a reference?

```suggestion
PbrShader& PbrShader::bindRoughnessTexture(Mn::GL::Texture2D& texture) {
```

Reply:
Simply because in `MaterialData` all the textures are stored in pointers, and thus can be passed into these functions immediately (I do not even need to judge if the pointer is nullptr in advance.)

Reply:
Commented because in Magnum all pointers that *can't* be null should be references and I think that's a policy we adopted here as well -- but not 100% sure. For consistency with other shaders it should be a reference tho ;)

On the contrary, in the `bindTextures()` call (or at least in the variant of it implemented in Phong), these *can* be `nullptr` -- and if it's so, given slot will not be bound. That's useful because in most cases you always have just a subset of the textures, almost never all of them **(In your case it's however not true, and if you pass a `nullptr` there, it'll crash.)**

Reply:
Why would it crash? 
If any nullptr is passed, it will be captured within the actual binding functions, such as bindRoughTextures etc.

Reply:
Ah, I overlooked the `if(texture)` part in there. Then it won't crash, you're right.